### PR TITLE
Tags create list show delete through open API and CLI

### PR DESCRIPTION
**Pull request by nopcoder**
API and CLI for tag list/create/delete/show

### CODE REVIEW

**Review by johnnyaug — CHANGES_REQUESTED** [review]

File: catalog/rocks/cataloger.go [hunk 1]
```diff
@@ -235,6 +236,69 @@ func (c *cataloger) ResetBranch(ctx context.Context, repository string, branch s
...
+		if len(tags) >= limit+1 {
+			break
+		}
+	}
+	if err := it.Err(); err != nil {
+		return nil, false, err
+	}
+	// return results (optional trimmed) and hasMore
+	hasMore := false
+	if len(tags) >= limit {
```
[the remark this answers is not in the document]
Comment (nopcoder) [reply catalog/rocks/cataloger.go:285]: 1. collecting limit+1 in the loop above
2. fixing this one - need to check > limit

File: cmd/lakectl/cmd/tag.go [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+		cmdutils.FuncValidator(0, uri.ValidateRefURI),
+	),
+	Run: func(cmd *cobra.Command, args []string) {
+		u := uri.Must(uri.Parse(args[0]))
+		client := getClient()
+		tag := args[1]
+		ctx := context.Background()
+		err := client.CreateTag(ctx, u.Repository, &models.Ref{
+			ID:       swag.String(u.Ref),
+			CommitID: swag.String(tag),
```
[the remark this answers is not in the document]
Comment (nopcoder) [reply cmd/lakectl/cmd/tag.go:86]: 1. The name `tag` is not good - will rename to commitRef. 
2. Let me know if you think we should change the API as the type 'ref' is used for input and output. while in the input we can accept short commit which we will resolve and in response, we will send the full commit.
3. The CLI takes the tag name from the uri parse, ex: lakefs://repo@v1.0 (tag v1.0)

[the remark this answers is not in the document]
Comment (johnnyaug) [reply cmd/lakectl/cmd/tag.go:86]: I understand. So maybe change `u` to tagUri

[the remark this answers is not in the document]
Comment (johnnyaug) [reply cmd/lakectl/cmd/tag.go:86]: I'm currently going over all lakectl API, for now just renaming is enough

File: cmd/lakectl/cmd/tag.go [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+		})
+		if err != nil {
+			DieErr(err)
+		}
+		Fmt("created tag '%s'\n", u.Ref)
```
[the remark this answers is not in the document]
Comment (nopcoder) [reply cmd/lakectl/cmd/tag.go:91]: 1. Fixed
2. Updated the api spec to support this one

File: catalog/rocks/fake_graveler_test.go [hunk 4]
```diff
@@ -360,3 +364,46 @@ func (m *FakeBranchIterator) Err() error {
 }
 
 func (m *FakeBranchIterator) Close() {}
+
+/// BARAK
```
Comment (johnnyaug) [inline catalog/rocks/fake_graveler_test.go:368]: nice touch

File: cmd/lakectl/cmd/tag.go [hunk 5]
```diff
@@ -70,25 +69,22 @@ var tagListCmd = &cobra.Command{
 }
 
 var tagCreateCmd = &cobra.Command{
-	Use:   "create <ref uri> <tag id>",
+	Use:   "create <ref uri> <commit ref>",
```
Comment (johnnyaug) [inline cmd/lakectl/cmd/tag.go:72]: the first one should be `<tag uri>`

**Review by johnnyaug — CHANGES_REQUESTED** [review]

File: catalog/rocks/fake_graveler_test.go (hunk 4 above)
Comment (nopcoder) [reply catalog/rocks/fake_graveler_test.go:368]: ha ha - forgot my text marker :)

**Review by johnnyaug — APPROVED** [review]

**Pull request merged**